pkg/ipam/allocator/multipool/pool_allocator.go

```diff
@@ -101,28 +102,170 @@ func (p *PoolAllocator) RestoreFinished() {
 	p.ready = true
 }
 
+func (p *PoolAllocator) addPool(poolName string, ipv4CIDRs []string, ipv4MaskSize int, ipv6CIDRs []string, ipv6MaskSize int) error {
+	v4, err := cidralloc.NewCIDRSets(false, ipv4CIDRs, ipv4MaskSize)
+	if err != nil {
+		return err
+	}
+
+	v6, err := cidralloc.NewCIDRSets(true, ipv6CIDRs, ipv6MaskSize)
+	if err != nil {
+		return err
+	}
+
+	p.pools[poolName] = cidrPool{
+		v4:         v4,
+		v6:         v6,
+		v4MaskSize: ipv4MaskSize,
+		v6MaskSize: ipv6MaskSize,
+	}
+
+	return nil
+}
+
 func (p *PoolAllocator) AddPool(poolName string, ipv4CIDRs []string, ipv4MaskSize int, ipv6CIDRs []string, ipv6MaskSize int) error {
-	if _, ok := p.pools[poolName]; ok {
+	if _, exists := p.pools[poolName]; exists {
 		return fmt.Errorf("pool %q already exists", poolName)
 	}
+	return p.addPool(poolName, ipv4CIDRs, ipv4MaskSize, ipv6CIDRs, ipv6MaskSize)
+}
 
-	v4, err := cidralloc.NewCIDRSets(false, ipv4CIDRs, ipv4MaskSize)
+func (p *PoolAllocator) UpsertPool(poolName string, ipv4CIDRs []string, ipv4MaskSize int, ipv6CIDRs []string, ipv6MaskSize int) error {
+	pool, exists := p.pools[poolName]
+	if !exists {
+		return p.addPool(poolName, ipv4CIDRs, ipv4MaskSize, ipv6CIDRs, ipv6MaskSize)
+	}
+
+	if ipv4MaskSize != pool.v4MaskSize {
+		return fmt.Errorf("cannot change IPv4 mask size in existing pool %q", poolName)
+	}
+	if ipv6MaskSize != pool.v6MaskSize {
+		return fmt.Errorf("cannot change IPv6 mask size in existing pool %q", poolName)
+	}
+
+	var newv4, newv6 []cidralloc.CIDRAllocator
+	var allocv4, allocv6 []string
+	var err error
+
+	// allocate new CIDR set for each IPv4 CIDR not yet in the pool
+	for _, v4cidr := range ipv4CIDRs {
+		prefix, parseErr := netip.ParsePrefix(v4cidr)
+		if parseErr != nil {
+			err = errors.Join(err, fmt.Errorf("invalid IPv4 CIDR: %w", parseErr))
+		}
+		if !hasCIDR(pool.v4, prefix) {
+			allocv4 = append(allocv4, v4cidr)
+		}
+	}
+	if err != nil {
+		return err
+	}
+	if len(allocv4) > 0 {
+		newv4, err = cidralloc.NewCIDRSets(false, allocv4, ipv4MaskSize)
+		if err != nil {
+			return err
+		}
+	}
+
+	// delete CIDR set for IPv4 CIDRs not present in the new IPv4 CIDRs
+	v4 := pool.v4
+	for i, oldv4cidr := range v4 {
+		exists := false
+		for _, v4cidr := range ipv4CIDRs {
+			_, cidr, parseErr := net.ParseCIDR(v4cidr)
+			if parseErr != nil {
+				err = errors.Join(err, fmt.Errorf("invalid IPv4 CIDR: %w", parseErr))
+			}
+			if oldv4cidr.IsClusterCIDR(cidr) {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			v4[i] = nil
+			v4 = slices.Delete(v4, i, i+1)
+		}
+	}
 	if err != nil {
 		return err
 	}
+	v4 = append(v4, newv4...)
 
-	v6, err := cidralloc.NewCIDRSets(true, ipv6CIDRs, ipv6MaskSize)
+	// allocate new CIDR set for each IPv6 CIDR not yet in the pool
```

Comment:
nit: How hard would it be to move the "detect new CIDRs and add them and remove old ones" logic into a separate function that we invoke once for IPv4 once for IPv6?

Reply:
Not too hard. Refactored into `(p *PoolAllocator).updateCIDRSets`, thanks for the suggestion!